afuc/afuc.h

```diff
@@ -84,9 +84,16 @@ typedef enum {
 	 * 0x0b22->0x0b24 (IB2).  Presumably $05 ends up w/ different value
 	 * for RB->IB1 vs IB1->IB2.
 	 */
-	OPC_CWRITE = 0x15,
-	OPC_CREAD  = 0x16,
-	OPC_OP17   = 0x17,  /* placeholder, seems a variant of cread?? */
+	OPC_CWRITE5 = 0x15,
+	OPC_CREAD5  = 0x16,
+
+	/* A6xx shuffled around the cwrite/cread opcodes and added new opcodes
+	 * that let you read/write directly to memory (and bypass the IOMMU?).
+	 */
```

Comment:
hmm, this explaination makes sequences like:
```
        14f1: b8030042  cread $03, [$00 + 0x042], 0x0
        14f2: b8040043  cread $04, [$00 + 0x043], 0x0
        14f3: b8050044  cread $05, [$00 + 0x044], 0x0
        14f4: a0434004  store $03, [$02 + 0x004], 0x4
        14f5: a0444004  store $04, [$02 + 0x004], 0x4
        14f6: a0454004  store $05, [$02 + 0x004], 0x4
```
look a bit funny, unless it is auto-incrementing `$02`.. although I suppose that could be what the 0x4 means..

I can't seem to find it atm, but at one point I had a patch that ported most of a5xx_debugfs.c for a6xx (I never got the reset thing to work properly without angering GMU).. but `REG_A6XX_CP_SQE_STAT_ADDR`/`_DATA` should get you a dump of the microcontroller registers (plus some extra stuff at the end)..  if you haven't realized that already.

Reply:
Yup, I'm pretty sure that 0x4 pre-increments the offset register. It comes in handy in combination with `(rep)`, for example:

```
        0a0b: 881c000b  mov $rem, 0x000b
        0a0c: 88020880  mov $02, 0x0880
        0a0d: a81c0026  cwrite $rem, [$00 + 0x026], 0x0
        0a0e: a8020027  cwrite $02, [$00 + 0x027], 0x0
        0a0f: a8c24001  cwrite $02, [$06 + 0x001], 0x4
        0a10: a8dc4001  cwrite $rem, [$06 + 0x001], 0x4
        0a11: acde4001  (rep)cwrite $addr2, [$06 + 0x001], 0x4
```

This copies 0xb dwords of data from some register `0x0880` to a series of control registers (there are a bunch of similar sequences nearby since this is some save/restore thing).

Reply:
Oh, and presumably this also handles incrementing the high 32 bits for `load/store` as well... so it's much simpler than manually advancing the pointer and re-programming the high bits, which is probably why it's done that way even without `(rep)`.

Reply:
Also, I should mention how I figured this out. I was looking at the firmware routines for reading/writing the preemption record which is partially documented in the downstream kernel driver at https://github.com/OnePlusOSS/android_kernel_oneplus_sm8150/blob/oneplus/SM8150_P_9.0/drivers/gpu/msm/adreno_a6xx.h#L46 and it uses these load/store opcodes. For example this part which loads the address from the register and compares the magic:

```
        0fe9: 88020004  mov $02, 0x0004
        0fea: 880308a3  mov $03, 0x08a3 ; CP_CONTEXT_SWITCH_PRIV_NON_SECURE_RESTORE_ADDR_LO
        0feb: a8020026  cwrite $02, [$00 + 0x026], 0x0 ; DWORDS
        0fec: a8030027  cwrite $03, [$00 + 0x027], 0x0 ; register load address
        0fed: 981e8006  mov $10, $addr2 ; get register
        0fee: 981e8806  mov $11, $addr2
        0fef: a8110058  cwrite $11, [$00 + 0x058], 0x0 ; set high bits of load/store
        ...
        0ff4: 8a03ae39  mov $03, 0xae39 << 16
        0ff5: 30639d6e  or $03, $03, 0x9d6e ; construct A6XX_CP_CTXRECORD_MAGIC_REF
        0ff6: b2020000  load $02, [$10 + 0x000], 0x0 ; load first dword
        0ff7: 98432010  cmp $04, $02, $03 ; compare with magic
        0ff8: c8800e2e  brne $04, b0, #l369 (#3630, 1e26)
```

Reply:
ok, that does look like a load, and it is using a 64b address.. although I do see a `load` near the start of the fw (ie. what it executes on power up before waiting for first packet) that does a load against $07 when $08 is never written..which seems odd.. presumably it does *something* valid since I don't see any branch/jump past it.  Possibly it is a bug and just works (as long as what it is loading is <32b) because $08 is zero at that point.  IIRC kgsl is using the same scheme we've been discussing for upstream (although without split pagetables) to place "kernel" gpu memory at low addresses, and per-context gpu memory above that.  (I don't believe load/store would be bypassing IOMMU.)

anyways, with the assembler it is easy enough to put in some of our own instructions and then an infinite loop (and then from kernel dump sqe registers when the kernel detects the hang).. that is how i worked out most of the instructions on a5xx.

Reply:
I was a little confused as well at some of the call-sites, but then I realized that the high bits are always written to control register 0x58 beforehand, so I think that register controls the high bits and it doesn't actually read two registers. See for example the snippet above.

Reply:
BTW, I'm not sure exactly the difference between this mechanism and the other one used by e.g. `CP_MEM_WRITE` involving `$addr` and `$data`, but I mentioned bypassing the IOMMU because this new method seems to only ever be used for kernel-created buffers like the preemption record and the SQE microcode itself. Although the `$addr` method is also used sometimes, sometimes mixing both in the same function!

Reply:
ok, so I have a theory.. the preemption records are allocated w/ `KGSL_MEMDESC_PRIVILEGED` (which should presumably mean they are not writeable from userspace cmdstream?), possibly `load`/`store` does privileged access?  I'll dig thru it a bit more as I have time (since I want to make sure I understand context switch to ensure there aren't any possible holes).  But not sure how much time I'll have in the next few days, since travel and family and such.